File: mycuration/src/test/java/com/phicdy/mycuration/presentation/presenter/RssListPresenterTest.java
package com.phicdy.mycuration.presentation.presenter;


import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;

import com.phicdy.mycuration.data.db.DatabaseAdapter;
import com.phicdy.mycuration.data.rss.Feed;
import com.phicdy.mycuration.domain.rss.UnreadCountManager;
import com.phicdy.mycuration.domain.task.NetworkTaskManager;
import com.phicdy.mycuration.presentation.view.RssListView;
import com.phicdy.mycuration.presentation.view.fragment.RssListFragment;
import com.phicdy.mycuration.util.PreferenceHelper;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

import java.util.ArrayList;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class RssListPresenterTest {

    private RssListPresenter presenter;
    private MockView view;

    private static final String FIRST_RSS_TITLE = "rss1";
    private static final String SECOND_RSS_TITLE = "rss2";
    private static final int FIRST_RSS_ID = 0;
    private static final int SECOND_RSS_ID = 1;

    private static final int FIRST_RSS_POSITION = 0;
    private static final int HIDE_OPTION_POSITION_WHEN_HIDDEN = 1;

    @Before
    public void setup() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        DatabaseAdapter.inject(adapter);
        ArrayList<Feed> allFeeds = new ArrayList<>();
        Feed firstRss = new Feed(FIRST_RSS_ID, FIRST_RSS_TITLE, "", "", "", 0, "");
        Feed secondRss = new Feed(SECOND_RSS_ID, SECOND_RSS_TITLE, "", "", "", 1, "");
        allFeeds.add(firstRss);
        allFeeds.add(secondRss);
        Mockito.when(adapter.getAllFeedsWithNumOfUnreadArticles()).thenReturn(allFeeds);
        NetworkTaskManager networkTaskManager = NetworkTaskManager.INSTANCE;

        Context mockContext = Mockito.mock(Context.class);
        Mockito.when(mockContext.getSharedPreferences("FilterPref", Context.MODE_PRIVATE))
                .thenReturn(Mockito.mock(SharedPreferences.class));
        PreferenceHelper.INSTANCE.setUp(mockContext);
        PreferenceHelper preferenceHelper = PreferenceHelper.INSTANCE;

        UnreadCountManager.INSTANCE.addFeed(firstRss);
        UnreadCountManager.INSTANCE.addFeed(secondRss);
        view = new MockView();
        presenter = new RssListPresenter(view, preferenceHelper, adapter, networkTaskManager, UnreadCountManager.INSTANCE);
    }

    @After
    public void tearDown() {
        UnreadCountManager.INSTANCE.clear();
    }

    @Test
    public void WhenFirstRssIsHiddenThenFirstEditTitleWillBeSecondRss() {
        // Default hidden option is enaled
        presenter.create();
        presenter.resume();
        presenter.onEditFeedMenuClicked(FIRST_RSS_POSITION);
        assertThat(view.editTitle, is(SECOND_RSS_TITLE));
    }

    @Test
    public void WhenRssIsNotHiddenThenFirstEditTitleWillBeFirstRss() {
        presenter.create();
        presenter.resume();
        // Disale hidden option
        presenter.onFeedListClicked(HIDE_OPTION_POSITION_WHEN_HIDDEN, new RssListFragment.OnFeedListFragmentListener() {
            @Override
            public void onListClicked(int feedId) {
            }

            @Override
            public void onAllUnreadClicked() {
            }
        });
        presenter.onEditFeedMenuClicked(FIRST_RSS_POSITION);
        assertThat(view.editTitle, is(FIRST_RSS_TITLE));
    }

    private class MockView implements RssListView {

        private String editTitle;

        @Override
        public void showDeleteFeedAlertDialog(int position) {

        }

        @Override
        public void showEditTitleDialog(int position, @NonNull String feedTitle) {
            this.editTitle = feedTitle;
        }

        @Override
        public void setRefreshing(boolean doScroll) {

        }

        @Override
        public void init(@NonNull ArrayList<Feed> feeds) {

        }

        @Override
        public void setTotalUnreadCount(int count) {

        }

        @Override
        public void onRefreshCompleted() {

        }

        @Override
        public void showEditFeedTitleEmptyErrorToast() {

        }

        @Override
        public void showEditFeedFailToast() {

        }

        @Override
        public void showEditFeedSuccessToast() {

        }

        @Override
        public void showDeleteSuccessToast() {

        }

        @Override
        public void showDeleteFailToast() {

        }

        @Override
        public void showAddFeedSuccessToast() {

        }

        @Override
        public void showGenericAddFeedErrorToast() {

        }

        @Override
        public void showInvalidUrlAddFeedErrorToast() {

        }

        @Override
        public void notifyDataSetChanged() {

        }

        @Override
        public void showAllUnreadView() {

        }

        @Override
        public void hideAllUnreadView() {

        }
    }
}

File: feature_rss_list/src/test/java/com/phicdy/mycuration/rss/RssListPresenterTest.kt
package com.phicdy.mycuration.rss


import android.content.Context
import android.content.SharedPreferences
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import com.phicdy.mycuration.data.preference.PreferenceHelper
import com.phicdy.mycuration.data.repository.RssRepository
import com.phicdy.mycuration.domain.task.NetworkTaskManager
import com.phicdy.mycuration.entity.Feed
import kotlinx.coroutines.runBlocking
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentMatchers.anyLong
import org.mockito.ArgumentMatchers.anyString

class RssListPresenterTest {

    private lateinit var presenter: RssListPresenter
    private val view = mock<RssListView>()
    private val mockPref = mock<SharedPreferences>()
    private val mockRssRepository = mock<RssRepository>()
    private val networkTaskManager = mock<NetworkTaskManager>()
    private lateinit var allFeeds: ArrayList<Feed>

    @Before
    fun setup() {
        // Mock two RSS returns
        val firstRss = Feed(FIRST_RSS_ID, FIRST_RSS_TITLE, "", Feed.DEDAULT_ICON_PATH, "", FIRST_RSS_UNREAD_COUNT, "")
        val secondRss = Feed(SECOND_RSS_ID, SECOND_RSS_TITLE, "", SECOND_RSS_ICON_PATH, "", SECOND_RSS_UNREAD_COUNT, "")
        allFeeds = arrayListOf(firstRss, secondRss)
        runBlocking {
            whenever(mockRssRepository.getAllFeedsWithNumOfUnreadArticles()).thenReturn(allFeeds)
            whenever(mockRssRepository.getNumOfRss()).thenReturn(2)
        }

        // Set up mock PreferenceHelper
        val mockContext = mock<Context>()
        whenever(mockContext.getSharedPreferences("FilterPref", Context.MODE_PRIVATE)).thenReturn(mockPref)
        PreferenceHelper.setUp(mockContext)
        val mockEdit = mock<SharedPreferences.Editor>()
        whenever(mockPref.edit()).thenReturn(mockEdit)
        whenever(mockEdit.putLong(any(), any())).thenReturn(mockEdit)

        presenter = RssListPresenter(view, PreferenceHelper, mockRssRepository, networkTaskManager)
    }

    @Test
    fun `when finish refresh then hide refresh view`() = runBlocking {
        presenter.onFinishUpdate()
        verify(view, times(1)).onRefreshCompleted()
    }

    @Test
    fun `when finish refresh then fetch RSS`() {
        runBlocking {
            presenter.onFinishUpdate()
            verify(mockRssRepository).getAllFeedsWithNumOfUnreadArticles()
        }
    }

    @Test
    fun `when finish refresh then reload RSS list`() {
        runBlocking { presenter.onFinishUpdate() }
        verify(view, times(1)).init(presenter.unreadOnlyFeeds.toRssListItem(RssListFooterState.UNREAD_ONLY))
    }

    @Test
    fun `when finish refresh then last update time will be updated`() {
        runBlocking { presenter.onFinishUpdate() }
        verify(mockPref.edit(), times(1)).putLong(anyString(), anyLong())
    }

    private fun ArrayList<Feed>.toRssListItem(state: RssListFooterState): List<RssListItem> = mutableListOf<RssListItem>().apply {
        add(RssListItem.All(this@toRssListItem.sumBy { it.unreadAriticlesCount }))
        add(RssListItem.Favroite)
        this@toRssListItem.map {
            this.add(RssListItem.Content(
                    rssId = it.id,
                    rssTitle = it.title,
                    rssIconPath = it.iconPath,
                    isDefaultIcon = it.iconPath.isBlank() || it.iconPath == Feed.DEDAULT_ICON_PATH,
                    unreadCount = it.unreadAriticlesCount
            ))
        }
        add(RssListItem.Footer(state))
    }

    companion object {

        private const val FIRST_RSS_TITLE = "rss1"
        private const val SECOND_RSS_TITLE = "rss2"
        private const val FIRST_RSS_ID = 0
        private const val SECOND_RSS_ID = 1
        private const val SECOND_RSS_ICON_PATH = "https://www.google.com/icon"
        private const val FIRST_RSS_UNREAD_COUNT = 0
        private const val SECOND_RSS_UNREAD_COUNT = 1
    }
}